contracts/access/AccessControlAdminRules.sol

```diff
@@ -0,0 +1,228 @@
+ * * The `DEFAULT_ADMIN_ROLE` 's admin can be only held by itself.
+ *
+ * Once you understand what the {constructor} parameters are, you can use this reference implementation:
+ *
+ * ```solidity
+ * contract MyToken is AccessControlAdminRules {
+ *   constructor() AccessControlAdminRules(
+ *     60 * 60 * 24, // 3 days
+ *     _msgSender() // Explicit initial `DEFAULT_ADMIN_ROLE` holder
+ *    ) {}
+ *}
+ * ```
+ *
+ * NOTE: The `delay` is only configurable in the constructor to avoid issues related with handling
+ * delay management during the transfer is pending to be completed.
+ *
+ * _Available since v4.9._
+ */
+abstract contract AccessControlAdminRules is IAccessControlAdminRules, IERC5313, AccessControl {
+    uint48 private immutable _delay;
+
+    address private _currentAdmin;
+    address private _pendingAdmin;
+
+    uint48 private _delayedUntil;
+
+    /**
+     * @dev Sets the initial values the internal delay and {admin}.
+     *
+     * The `delay` value is immutable. It can only be set at the constructor.
+     */
+    constructor(uint48 delay, address initialAdmin) {
+        _delay = delay;
+        _grantRole(DEFAULT_ADMIN_ROLE, initialAdmin);
+    }
+
+    /**
+     * @dev See {IERC5313-owner}
+     */
+    function owner() public view virtual returns (address) {
+        return admin();
+    }
+
+    /**
+     * @dev See {IAccessControlAdminRules-admin}
+     */
+    function admin() public view virtual returns (address) {
+        return _currentAdmin;
+    }
+
+    /**
+     * @dev See {IAccessControlAdminRules-delayedUntil}
+     */
+    function delayedUntil() public view virtual returns (uint48) {
+        return _delayedUntil;
+    }
+
+    /**
+     * @dev See {IAccessControlAdminRules-pendingAdmin}
+     */
+    function pendingAdmin() public view virtual returns (address) {
+        return _pendingAdmin;
+    }
+
+    /**
+     * @dev See {IERC165-supportsInterface}.
+     */
+    function supportsInterface(bytes4 interfaceId) public view virtual override returns (bool) {
+        return interfaceId == type(IAccessControlAdminRules).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    /**
+     * @dev See {IAccessControlAdminRules-beginAdminTransfer}
+     */
+    function beginAdminTransfer(address newAdmin) public virtual override onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(delayedUntil() == 0, "AccessControl: pending admin already set");
+        _delayedUntil = SafeCast.toUint48(block.timestamp) + _delay;
+        _pendingAdmin = newAdmin;
+        emit AdminRoleChangeStarted(pendingAdmin(), delayedUntil());
+    }
+
+    /**
+     * @dev See {IAccessControlAdminRules-acceptAdminTransfer}
+     */
+    function acceptAdminTransfer() public virtual {
+        address pendingAdminOwner = pendingAdmin();
+        require(
+            _hasAdminTransferDelayPassed() && _msgSender() == pendingAdminOwner,
+            "AccessControl: delay must be met and caller must be pending admin"
```

Comment:
String revert reasons should be shorter to avoid taking up too much bytecode.

Reply:
I'm ok to keep the `cancelAdminTransfer` for workflow were the cancel is not preceding a `beginAdminTransfer`.

Said otherwize: yes, lets keep the cancel function, but lets also remove the `require(delayedUntil() == 0, "AccessControl: pending admin already set");` check in `beginAdminTransfer`

Reply:
for the revert reason, why not use custom errors ?

Reply:
It also feels odd to me to introduce them right now. I think we can safely wait until 5.0.

> I'm ok to keep the ...

Will follow in the other comments thread.